hackathon_site/hardware/serializers.py

```diff
@@ -46,9 +50,91 @@ def get_unique_hardware_count(obj: Category):
+
+    team_id = serializers.PrimaryKeyRelatedField(
+        queryset=TeamEvent.objects.all(), many=False, required=True
+    )
+    hardware = OrderCreateHardwareSerializer(many=True, required=True)
+
+    @staticmethod
+    def merge_requests(hardware_requests):
+        return functools.reduce(
+            lambda x, y: x + y,
+            [Counter({e["id"]: e["quantity"]}) for e in hardware_requests],
+        )
+
+    def create(self, validated_data):
+        requested_hardware = self.merge_requests(
+            hardware_requests=validated_data["hardware"]
+        )
+        relevant_past_order_items = (
+            OrderItem.objects.filter(hardware__in=requested_hardware.keys(),)
+            .exclude(order__status="Cancelled")
+            .select_related("hardware")
+        )
+        team_past_hardware_counts = Counter(
+            (
+                map(
+                    lambda item: item.hardware,
+                    relevant_past_order_items.filter(
+                        order__team=validated_data["team_id"],
+                    ).all(),
+                )
+            )
+        )
+        all_past_hardware_counts = Counter(
+            (map(lambda item: item.hardware, relevant_past_order_items.all(),))
+        )
+        new_order = None
+        unfulfilled_hardware_requests = dict()
+        for (hardware, requested_quantity) in requested_hardware.items():
+            allowed_quantity = min(
+                hardware.max_per_team - team_past_hardware_counts[hardware],
+                hardware.quantity_available - all_past_hardware_counts[hardware],
+            )
+            if allowed_quantity <= 0:
+                unfulfilled_hardware_requests[hardware] = requested_quantity
+                continue
+            if new_order is None:
+                new_order = Order.objects.create(
+                    team=validated_data["team_id"], status="Submitted"
+                )
+            if allowed_quantity < requested_quantity:
+                unfulfilled_hardware_requests[hardware] = (
+                    requested_quantity - allowed_quantity
+                )
+            for _ in itertools.repeat(None, min(allowed_quantity, requested_quantity)):
+                OrderItem.objects.create(
+                    order=new_order, hardware=hardware,
+                )
+
+        return new_order, unfulfilled_hardware_requests
+
+
+class OrderCreateResponseSerializer(serializers.Serializer):
+    def to_representation(self, instance):
+        response = {"order_id": instance.id}
+        fulfilled_order_items = instance.items
+        fulfilled_count = Counter(
+            map(lambda item: item.hardware_id, fulfilled_order_items.all())
+        )
+        response["hardware"] = dict(fulfilled_count)
+        unfulfilled = self.context["unfulfilled"]
+        response["errors"] = {
+            hardware.id: count for (hardware, count) in unfulfilled.items()
+        }
```

Comment:
i thought it is enough for the server to return the pk and unfulfilled amount for each hardware. The front end can figure out how to formate the response to present to the user (unless it is difficult in react to do that...?)

Reply:
I would do that on the backend instead of the frontend, since that's what most of our other APIs do. Having the severity key also lets us distinguish between warnings (eg ran out of some hardware) vs errors (invalid hardware id, constraints not met)